core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -983,8 +979,14 @@ protected void onException(Channel channel, Exception e) throws IOException {
      */
     protected abstract void closeChannels(List<Channel> channel) throws IOException;
 
-
-    protected abstract void sendMessage(Channel channel, BytesReference reference, Runnable sendListener) throws IOException;
+    /**
+     * Sends message to channel. The listener will be called with an exception if the send operation throws an exception. If the operation
+     * is successful, the listener will be called with null.
```

Comment:
Although I have not extensively explored number 3. Taking a quick look, it might be a little more challenging to figure out what a write is in the pipeline as we only send a bytes reference to netty. So figuring out if a write is a request, response, or ping might be tricky.

I guess in this PR I'm fine with the ActionListener approach if you guys think that is the correct approach.